### DISCUSSION TITLE
VTextField validation

### DISCUSSION THREAD

**Discussion by jarrah42** [Kitware/trame#517] [Q&A]
The only example of `VTextField` validation I can find is in the `09_advanced/custom_js` example where a custom JS function is used. Is it possible to use a python function/expression for validation? I've tried setting rules to `("[self.validate]",)` but it complains about an undefined object.

**Comment by jourdain** [comment]
Indeed `self.validate` is not available on the client side. But if the function can be async, you can connect it to a server one.

**Comment by jarrah42** [comment]
Would you have an example of how to do that you could share?

**Comment by jourdain** [comment]
Not right now, as I would have to test it. But the idea is to call the JS trigger function like shown in the doc [here](https://trame.readthedocs.io/en/latest/client.intro.html#trigger)

**Comment by jarrah42** [comment]
Ok thanks, I'll see if I can get it to work.

**Comment by jarrah42** [comment]
It's possible to trigger the function, but not get the value that is passed to be validated. Any thoughts on how to do that?

I was also wondering why does click allow a python function to be supplied and not rules? Is something happening behind the scenes to make this work?

**Reply by jourdain** [reply]
What did you wrote? It is always easier to fix a mistake than coming up with the full code. 

The difference between click and rules is that click is an event while rules is a property. For event we can easily write the JS for you when you pass a Python function. But for a property we can't assume the format of what JS expect. So it is up to you to write the JavaScript to match what the component expect. You can definitely make call to the server with arguments, but now you start to need to know more about JS than before... Trame can't guess it for you. You just need to make the call correctly.

**Comment by jarrah42** [comment]
I used:

```
vuetify.VTextField(v_model=("dmin", 0.7), label="d(min)", suffix="A", density="compact", type="text", 
                                variant="outlined", 
                                hide_spin_buttons="true", hide_details="true",
                                rules=("[trigger('validate_dmin')]",), validate_on="input")
```

with the trigger function:

```
@self.ctrl.trigger("validate_dmin")
def validate_dmin(*args, **kwargs) -> bool:
      print("args=", args)
      print("kwargs=", kwargs)
```

**Reply by jourdain** [reply]
```python
rules=("[(v) => trigger('validate_dmin', [v])]",)
```

**Comment by jarrah42** [comment]
Thanks!
